compiler/rustc_codegen_llvm/src/llvm_util.rs

```diff
@@ -308,30 +277,59 @@ pub fn check_tied_features(
 /// Used to generate cfg variables and apply features
 /// Must express features in the way Rust understands them
 pub fn target_features(sess: &Session, allow_unstable: bool) -> Vec<Symbol> {
-    let target_machine = create_informational_target_machine(sess);
+    let mut features = FxHashSet::default();
+
+    // Add base features for the target
+    let target_machine = create_informational_target_machine(sess, true);
+    features.extend(
+        sess.target
+            .supported_target_features()
+            .iter()
+            .filter(|(feature, _, _)| {
+                // skip checking special features, as LLVM may not understands them
+                if RUSTC_SPECIAL_FEATURES.contains(feature) {
+                    return true;
+                }
+                // check that all features in a given smallvec are enabled
+                for llvm_feature in to_llvm_features(sess, feature) {
+                    let cstr = SmallCStr::new(llvm_feature);
+                    if !unsafe { llvm::LLVMRustHasFeature(&target_machine, cstr.as_ptr()) } {
+                        return false;
+                    }
+                }
+                true
+            })
+            .map(|(feature, _, _)| Symbol::intern(feature)),
+    );
+
+    // Add enabled features
+    for (enabled, feature) in
+        sess.opts.cg.target_feature.split(',').filter_map(|s| match s.chars().next() {
+            Some('+') => Some((true, Symbol::intern(&s[1..]))),
+            Some('-') => Some((false, Symbol::intern(&s[1..]))),
+            _ => None,
+        })
+    {
+        if enabled {
+            features.extend(sess.target.implied_target_features(std::iter::once(feature)));
+        } else {
+            features.remove(&feature);
```

Comment:
We want to do the reverse here as well: remove all features that depend on the removed feature.

Reply:
Yeah, that's why I didn't do it.  My implementation slightly differs for `+avx2,-avx`, which does nothing currently but enables `avx2` without `avx` with my changes.  Neither really makes sense to me, so I'm not sure which is appropriate.  I could match the current behavior by searching through previously enabled features upon disabling a feature, I guess I should make that change?

Reply:
Reading @Amanieu's request more precisely I think it's actually correct, features that depend on the removed feature should be removed, but not the other way around.

Reply:
I think I have this properly implemented now.